### PR TITLE
Adding a `ReFT` notebook to the tutorials section

### PR DESCRIPTION
**Pull request by julian-fong**
This PR aims to add a tutorial notebook to utilize the `Loreft` adapter to fine-tune `roberta-base` on the `mnli` dataset.

Reviews appreciated!

### CODE REVIEW

File: notebooks/ReFT_Adapters_Finetuning.ipynb [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Fine Tuning using ReFT Adapters\n",
+    "\n",
+    "In this tutorial, we will be demonstrating how to fine-tune a language model using [Representation Finetuning for Language Models](https://arxiv.org/abs/2404.03592)\n",
+    "\n",
+    "We will use a traditional large language model and focus on fine tuning via ReFT adapters rather than the traditional full model fine tuning.\n",
```
Comment (calpt) [inline notebooks/ReFT_Adapters_Finetuning.ipynb:11]: i'm not sure "traditional large language model" is a good description here. maybe something like "lightweight encoder model" is a good description for Roberta nowadays.

File: notebooks/ReFT_Adapters_Finetuning.ipynb [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+    "data_collator = default_data_collator"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Model and Adapter initialization\n",
+    "\n",
+    "We load the `roberta-base` model along with the `LoReftConfig`. We can initalize a `reft` config with only one line of code, and can add it to our base model using the `add_adapter` function. On top of that, we can add a classification head to our adapter specifying 3 labels.\n",
```
Comment (calpt) [inline notebooks/ReFT_Adapters_Finetuning.ipynb:165]: could be nice to link our docs on ReFT here: https://docs.adapterhub.ml/methods.html#reft and mention that there is explanation on supported config parameters.

File: notebooks/ReFT_Adapters_Finetuning.ipynb [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    "from adapters import AdapterTrainer\n",
+    "from transformers import TrainingArguments, Trainer\n",
+    "\n",
+    "training_args = TrainingArguments(\n",
+    "    output_dir='./results',\n",
+    "    eval_strategy='epoch',\n",
+    "    learning_rate=6e-4,\n",
+    "    per_device_train_batch_size=32,\n",
+    "    per_device_eval_batch_size=32,\n",
+    "    num_train_epochs=2,\n",
```
Comment (calpt) [inline notebooks/ReFT_Adapters_Finetuning.ipynb:266]: please make a note that usually you'd train longer and this is only for demo purposes.

File: notebooks/ReFT_Adapters_Finetuning.ipynb [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    "mapping = {\n",
+    "    0: \"neutral\",\n",
+    "    1: \"entailment\",\n",
+    "    2: \"contradiction\"\n",
+    "}\n",
+    "\n",
+    "def infer_text(text):\n",
+    "    input_ids = tokenizer(text, truncation=True, padding='max_length')\n",
+    "    input_ids[\"input_ids\"] = torch.tensor(input_ids[\"input_ids\"])\n",
+    "    input_ids[\"attention_mask\"] = torch.tensor(input_ids[\"attention_mask\"])\n",
```
Comment (calpt) [inline notebooks/ReFT_Adapters_Finetuning.ipynb:384]: can add `return_tensors="pt"` (same as in preprocess_function) to tokenizer call so you don't need to convert to tensors afterwards

**Review by calpt — COMMENTED** [review]
Looks great overall, thanks for working on this! Left some mostly minor comments to be adressed before we merge :)

Additionally, could you add this new notebook to the list in the README of the notebooks folder, thanks!

File: notebooks/ReFT_Adapters_Finetuning.ipynb (hunk 4 above)
Comment (julian-fong) [reply notebooks/ReFT_Adapters_Finetuning.ipynb:384]: ah, good catch!

**Comment by julian-fong** [comment]
> Looks great overall, thanks for working on this! Left some mostly minor comments to be adressed before we merge :)
> 
> Additionally, could you add this new notebook to the list in the README of the notebooks folder, thanks!

Sounds good, should the whisper notebook be added as well? I dont see it in the read.me

**Comment by TimoImhof** [comment]
> > Looks great overall, thanks for working on this! Left some mostly minor comments to be adressed before we merge :)
> > Additionally, could you add this new notebook to the list in the README of the notebooks folder, thanks!
> 
> Sounds good, should the whisper notebook be added as well? I dont see it in the read.me

Yes please, we forgot that 👍

**Review by calpt — APPROVED** [review]
thanks for including the feedback!

**Pull request merged**